lib/puma/request.rb

```diff
@@ -73,8 +73,13 @@ def handle_request(client, lines, requests)
 
       begin
         begin
-          status, headers, res_body = @thread_pool.with_force_shutdown do
-            @app.call(env)
+          if env[REQUEST_METHOD] != CONNECT
+            status, headers, res_body = @thread_pool.with_force_shutdown do
+              @app.call(env)
+            end
+          else
+            @log_writer.log "CONNECT method used, but it's not supported by puma."
```

Comment:
Actually, the [text on MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/501) suggests a little different logic is possible:

```ruby 
if ALLOWED_HTTP_METHODS.include?(env[REQUEST_METHOD])
  # call app 
else 
  # return 501
end
```